olp-cpp-sdk-core/include/olp/core/thread/TaskContinuation.h

```diff
@@ -0,0 +1,397 @@
+struct TypedSmartPointer : UntypedSmartPointer {
+  explicit TypedSmartPointer(SmartPointer&& ptr) : m_ptr(std::move(ptr)) {}
+
+  void* get() const override { return static_cast<void*>(m_ptr.get()); }
+
+  SmartPointer m_ptr;
+};
+
+template <typename SmartPointer>
+inline std::unique_ptr<UntypedSmartPointer> make_untyped_unique(
+    SmartPointer&& ptr) {
+  return std::make_unique<TypedSmartPointer<SmartPointer>>(std::move(ptr));
+}
+
+/**
+ * \brief The ContinuationImpl class.
+ */
+class ContinuationImpl {
+ public:
+  using OutResult = std::unique_ptr<UntypedSmartPointer>;
+
+  using Task = std::function<OutResult(void*)>;
+  using Callback = std::function<void(void*)>;
+  using AsyncTask = std::function<void(void*, Callback)>;
+  using FinalCallback = std::function<void(void*, bool)>;
+
+  ContinuationImpl() {}
+
+  /// Constructs a continuation using the given async task
+  ContinuationImpl(std::shared_ptr<TaskScheduler> group,
+                   std::shared_ptr<ExecutionContext> context,
+                   std::pair<AsyncTask, Task>&& async);
+
+  ContinuationImpl Then(std::pair<AsyncTask, Task>&& async);
+
+  void Run(FinalCallback callback) const;
+
+  std::shared_ptr<ExecutionContext> GetContext() const;
+
+ private:
+  struct Processor;
+
+  using SharedStep = std::shared_ptr<const std::pair<AsyncTask, Task>>;
+
+  std::shared_ptr<olp::thread::TaskScheduler> m_scheduler;
+  std::shared_ptr<ExecutionContext> m_execution_context;
+  std::deque<std::pair<AsyncTask, Task>> deque_;
+};
+
+}  // namespace internal
+
+/// Template specialization for Continuation of void type. It has the same
+/// interface as the generic version.
+template <>
+class Continuation<void> {
+  template <typename OtherResultType>
+  friend class Continuation;
+
+ public:
+  using Result = void;
+
+  template <typename Callable>
+  Continuation<internal::AsyncResultType<Callable>> Then(Callable callable) {
+    using NewResultType = internal::AsyncResultType<Callable>;
+    using Function = internal::TypeToFunctionInput<NewResultType>;
+    return Then(std::function<void(Function)>(std::move(callable)));
+  }
+
+  Continuation() = delete;
+
+  template <typename NewType>
+  Continuation<internal::RemoveRefAndConst<NewType>> Then(
+      std::function<void(std::function<void(NewType)>)> async) {
+    return m_cont.Then(toAsyncTask({}, std::move(async)));
+  }
+
+  Continuation Then(std::function<void(std::function<void()>)> async);
+
+  using ContinuationImpl = internal::ContinuationImpl;
+  using OutResult = ContinuationImpl::OutResult;
+  using Task = ContinuationImpl::Task;
+  using Callback = ContinuationImpl::Callback;
+
+ private:
+  template <typename NewType>
+  static std::pair<ContinuationImpl::AsyncTask, ContinuationImpl::Task>
+  toAsyncTask(std::shared_ptr<ExecutionContext> context,
+              std::function<void(std::shared_ptr<ExecutionContext>,
+                                 std::function<void(NewType)>)>&& async) {
```

Comment:
Make alias for execution func type:
```cpp
template <typename InType>
using ExecuteFuncType = std::function<void(ExecutionContext, std::function<void(InType)>)>;
```

considering that ExecutionContext is a shared pimpl already, etc.

Reply:
can't create alias, as InType has to be used afterwards.